src/mlpack/methods/ann/vanilla_rbm.hpp

```diff
@@ -0,0 +1,189 @@
+/**
+ * @file vanilla_rbm.hpp
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_VANILLA_RBM_HPP
+#define MLPACK_METHODS_ANN_VANILLA_RBM_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/math/random.hpp>
+#include "layer/layer.hpp"
+#include "layer/base_layer.hpp"
+#include "visitor/weight_set_visitor.hpp"
+#include "visitor/forward_visitor.hpp"
+#include "visitor/weight_size_visitor.hpp"
+#include "activation_functions/softplus_function.hpp"
+#include "init_rules/gaussian_init.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InitializationRuleType,
+         typename VisibleLayerType,
+         typename HiddenLayerType>
+class VanillaRBM
```

Comment:
I guess if you pass layer types as template parameters these parameters completely define the type of the RBM. So, for example `RBM<... , BinaryRBMLayer, BinaryRBMLayer>` represents the binary RBM and `RBM<..., SSRBMVisibleLayer, SSRBMHiddenLayer>` represents the spike and slab RBM. In such a way it is better to call the class just `RBM`.